File: ios/InverterMonitor/InverterMonitor/Views/Live/SystemInfoGrid.swift
import SwiftUI

struct SystemInfoGrid: View {
    @ObservedObject var live: LiveDashboardViewModel

    private let columns = [GridItem(.adaptive(minimum: 140), spacing: 12)]

    var body: some View {
        LazyVGrid(columns: columns, spacing: 12) {
            infoTile(
                title: "Inverter Temp",
                value: formatTemp(live.status.system.temperature),
                unit: "°C",
                footnote: "Today peak \(formatTemp(live.summary.temperatureMax))°C"
            )
            infoTile(
                title: "DC Bus Voltage",
                value: formatBus(live.status.system.busVoltage),
                unit: "V",
                footnote: "Mode: \(live.status.system.modeLabel)"
            )
            infoTile(
                title: "Connection",
                value: live.connection.label,
                unit: nil,
                footnote: lastUpdateText
            )
            infoTile(
                title: "Reading Cycle",
                value: formatDuration(live.status.timing?.durationMs),
                unit: "ms",
                footnote: "Errors \(errorCount) / \(totalReadings)"
            )
        }
    }

    private func infoTile(title: String, value: String, unit: String?, footnote: String) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            Text(title)
                .font(.caption.weight(.semibold))
                .foregroundStyle(Palette.subtleText)
            HStack(alignment: .lastTextBaseline, spacing: 4) {
                Text(value)
                    .font(.system(size: 18, weight: .bold, design: .rounded))
                    .foregroundStyle(.white)
                    .monospacedDigit()
                    .contentTransition(.numericText(value: Double(value) ?? 0))
                    .animation(.easeOut(duration: 0.3), value: value)
                if let unit {
                    Text(unit)
                        .font(.caption)
                        .foregroundStyle(Palette.subtleText)
                }
            }
            Text(footnote)
                .font(.caption2)
                .foregroundStyle(Palette.subtleText)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(12)
        .card()
        .accessibilityElement(children: .combine)
        .accessibilityLabel("\(title) \(value)\(unit.map { " \($0)" } ?? ""). \(footnote)")
    }

    private var lastUpdateText: String {
        guard let last = live.lastUpdate else { return "Waiting…" }
        return "Last update \(Self.timeFormatter.string(from: last))"
    }

    private var errorCount: Int { live.readingStats?.errorCount ?? 0 }
    private var totalReadings: Int { live.readingStats?.totalReadings ?? 0 }

    private func formatTemp(_ t: Double) -> String {
        if t <= 0 { return "—" }
        return "\(Int(t.rounded()))"
    }

    private func formatBus(_ v: Double) -> String {
        if v <= 0 { return "—" }
        return "\(Int(v.rounded()))"
    }

    private func formatDuration(_ ms: Double?) -> String {
        guard let ms, ms > 0 else { return "—" }
        return "\(Int(ms.rounded()))"
    }

    private static let timeFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "HH:mm:ss"
        return f
    }()
}

File: android/InverterMonitor/app/src/main/java/com/bilalahmad/invertermonitor/ui/screens/live/SystemInfoGrid.kt
package com.bilalahmad.invertermonitor.ui.screens.live

import androidx.compose.foundation.layout.*
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.bilalahmad.invertermonitor.ui.components.card
import com.bilalahmad.invertermonitor.ui.theme.Palette
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import kotlin.math.roundToInt

@Composable
fun SystemInfoGrid(
    temperature: Double,
    busVoltage: Double,
    modeLabel: String,
    connectionLabel: String,
    lastUpdate: Date?,
    todayTemperatureMax: Double,
    readingDurationMs: Double?,
    errorCount: Int,
    totalReadings: Int,
) {
    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
            InfoTile(
                title = "Inverter Temp",
                value = if (temperature <= 0) "—" else "${temperature.roundToInt()}",
                unit = "°C",
                footnote = "Today peak ${todayTemperatureMax.roundToInt()}°C",
                modifier = Modifier.weight(1f),
            )
            InfoTile(
                title = "DC Bus Voltage",
                value = if (busVoltage <= 0) "—" else "${busVoltage.roundToInt()}",
                unit = "V",
                footnote = "Mode: $modeLabel",
                modifier = Modifier.weight(1f),
            )
        }
        Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
            InfoTile(
                title = "Connection",
                value = connectionLabel,
                unit = null,
                footnote = lastUpdate?.let { "Last update ${SimpleDateFormat("HH:mm:ss", Locale.getDefault()).format(it)}" } ?: "Waiting…",
                modifier = Modifier.weight(1f),
            )
            InfoTile(
                title = "Reading Cycle",
                value = readingDurationMs?.takeIf { it > 0 }?.roundToInt()?.toString() ?: "—",
                unit = "ms",
                footnote = "Errors $errorCount / $totalReadings",
                modifier = Modifier.weight(1f),
            )
        }
    }
}

@Composable
private fun InfoTile(title: String, value: String, unit: String?, footnote: String, modifier: Modifier = Modifier) {
    Column(
        modifier = modifier
            .card()
            .padding(12.dp),
        verticalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        Text(title, color = Palette.SubtleText, fontSize = 11.sp, fontWeight = FontWeight.SemiBold)
        Row(verticalAlignment = Alignment.Bottom) {
            Text(value, color = Color.White, fontSize = 17.sp, fontWeight = FontWeight.Bold)
            if (unit != null) {
                Spacer(Modifier.width(4.dp))
                Text(unit, color = Palette.SubtleText, fontSize = 11.sp)
            }
        }
        Text(footnote, color = Palette.SubtleText, fontSize = 10.sp)
    }
}
